From b47179241127b59d45810b012b1848cefbd26ad4 Mon Sep 17 00:00:00 2001
From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Tue, 23 Sep 2014 20:23:33 -0400
Subject: [virt] kvm: nested vmx - disable perf cpuid reporting

Message-id: <20140923202333.GA32645@amt.cnet>
Patchwork-id: 96506
O-Subject: [RHEL7.1 PATCH] KVM: nested VMX: disable perf cpuid reporting
Bugzilla: 1083969
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>

commit 20338adcdd956ad444b97315d21b1492725d7fce upstream of kvm.git/queue branch

Initilization of L2 guest with -cpu host, on L1 guest with -cpu host
triggers:

(qemu) KVM: entry failed, hardware error 0x7
...
nested_vmx_run: VMCS MSR_{LOAD,STORE} unsupported

Nested VMX MSR load/store support is not sufficient to
allow perf for L2 guest.

Until properly fixed, trap CPUID and disable function 0xA.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
BZ: 1083969
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index 38a0afe..65dc03e 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -767,6 +767,12 @@ void kvm_cpuid(struct kvm_vcpu *vcpu, u32 *eax, u32 *ebx, u32 *ecx, u32 *edx)
  if (!best)
   best = check_cpuid_limit(vcpu, function, index);
 
+ /*
+  * Perfmon not yet supported for L2 guest.
+  */
+ if (is_guest_mode(vcpu) && function == 0xa)
+  best = NULL;
+
  if (best) {
   *eax = best->eax;
   *ebx = best->ebx;
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 801332e..0fd49f7 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -6871,6 +6871,8 @@ static bool nested_vmx_exit_handled(struct kvm_vcpu *vcpu)
  case EXIT_REASON_TASK_SWITCH:
   return 1;
  case EXIT_REASON_CPUID:
+  if (kvm_register_read(vcpu, VCPU_REGS_RAX) == 0xa)
+   return 0;
   return 1;
  case EXIT_REASON_HLT:
   return nested_cpu_has(vmcs12, CPU_BASED_HLT_EXITING);
-- 
1.7.1